fix(animations): allow animations with unsupported CSS properties

dario-piotrowicz opened the pull request:
currently animations with unsupported CSS properties cause a hard error
and the crash of the animation itself, instead of this behaviour just
ignore such properties and provide a warning for the developer in
the console (only in dev mode)

resolves #23195

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## Issue

Issue Number: #23195



## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

I don't think it's a breaking change, unless someone was actually in some circumstances relying on the animations crushing when finding unsupported CSS properties :thinking: 

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

You can notice that the warnings I've added are `string`s which is inconsistent with the errors being `any`s, I believe the errors should actually also be strings and I've already opened #44726 to update their type :slightly_smiling_face:

jessicajaniuk commented on packages/animations/browser/src/dsl/animation_ast_builder.ts:
```diff
@@ -307,8 +314,8 @@ export class AnimationAstBuilderVisitor implements AnimationDslVisitor {
         if (!this._driver.validateStyleProperty(prop)) {
-          context.errors.push(`The provided animation property "${
-              prop}" is not a supported CSS property for animations`);
+          delete tuple[prop];
```
Deleting properties has a permanent negative impact on performance. Do we need to delete this here?

dario-piotrowicz replied:
Yes I thought the same! :slightly_smiling_face: 

I did actually try not to remove the property and all still worked the same, I thought that the cleanest thing was to just remove the property so that we actually end up with the exact same result as if the property wasn't there in the first place, and figured the negative impact would be negligible

I can definitely remove the `delete` here, but I'll need to investigate things first, I think that something somewhere may still try to use it (and fail), and if that's the case, it could cause unexpected behaviors in some cases or even have a worst performance impact compared to just use the `delete`

Moreover I was thinking that when you land your `Map`s PR this wouldn't be an object `delete` anymore but a Map one which should be totally fine :slightly_smiling_face: 

Anyways let me do a bit of investigation and see if I can figure the impact of keeping the prop lingering :slightly_smiling_face::+1:

dario-piotrowicz replied:
I just checked and the property if not deleted/removed actually gets treated like any other and ends up being applied to the element (of course it does not produce any animation/visual effect since it is not supported anyways):
![Screenshot at 2022-01-22 15-27-33](https://user-images.githubusercontent.com/61631103/150644867-5df8060c-e767-4943-ba2e-8e5911f19d52.png)

this means (I also double checked it) that it passes through params interpolation, prop normalization, and all the various (rather small) steps that each property passes through...

On the other hand the delete I'm adding only occurs once per property (during the trigger initialization), whilst the aforementioned operation occur for any transition/animation (/when necessary)

So I am not actually sure what would be the best solution, what do you think? :thinking: 

By the way, one solution could also be to remove the `delete` and add a Todo comment referring to your Maps PR (#44482) and saying that when that one lands a delete should be added here (as it would be quite more efficient at that point, right? :thinking:)

jessicajaniuk replied:
If it is passing a field we don't want, then I suggest a hybrid solution. I'd say keep the delete statement for now, but add a TODO on that line for me. It would be great if you could also add a comment referring to this change on my PR so I remember to address it before v14.

dario-piotrowicz replied:
of course! :slightly_smiling_face::+1:

jessicajaniuk reviewed (commented):
Thanks, @dario-piotrowicz! I like this a lot. Definitely take a look at the comment below. One thing I've been thinking about as I read this is that there's similar code used every time you output them to `console.warn`. This suggests to me maybe an abstraction makes sense. What do you think?

dario-piotrowicz commented:
> Thanks, @dario-piotrowicz! I like this a lot. Definitely take a look at the comment below. One thing I've been thinking about as I read this is that there's similar code used every time you output them to `console.warn`. This suggests to me maybe an abstraction makes sense. What do you think?

Thanks! :smiley: 

Yes an abstraction would make total sense, the issue is that if you look at the `console.warn`s they are not all the same, I am actually following what the errors do (before each of my `console.warn`s you can see an error being thrown and that I am generating the warning string in the same manner as the error throw).

I am not sure why they are different and figured the safest thing would be to follow that for the introduction of warnings, but I could definitely abstract the console warnings in some reusable way and then we can pick the errors up separately?

or would it make sense to just keep the warnings as they are for this PR and abstract both errors and warnings at the same time in a separate PR? :thinking: 

regardless there are many ways these errors can pop up, I haven't really tried triggering them all, I will have to have an investigation in this too to see if there is a reason we have different formats for errors in different places or not (probably not, but still would be good to have a look around I think :smile:)

dario-piotrowicz commented:
@jessicajaniuk I've looked into the errors and they are all just results of `buildAnimationAst` calls, so it doesn't really look like there is any reason to present them differently (on the contrary, it would be good to display them in a consistent format)

So I'll add a commit to add some shared functionality for the errors and warnings :slightly_smiling_face::+1:

dario-piotrowicz commented on packages/animations/browser/src/render/transition_animation_engine.ts:
```diff
@@ -1101,7 +1100,7 @@ export class TransitionAnimationEngine {
       erroneousTransitions.forEach(instruction => {
         errors.push(`@${instruction.triggerName} has failed due to:\n`);
-        instruction.errors!.forEach(error => errors.push(`- ${error}\n`));
+        instruction.errors!.forEach(error => errors.push(error));
```
I needed to change this because with the new change it was looking pretty awkward:
![Screenshot at 2022-01-22 19-07-47](https://user-images.githubusercontent.com/61631103/150654614-f38f8452-b44c-4855-8161-b98c6b56ba3b.png)
(note the double `-`s)

without the `-` here it gets a little better:
![Screenshot at 2022-01-22 19-31-46](https://user-images.githubusercontent.com/61631103/150654626-c936ec7d-5b41-4ac2-b6d8-7fa4b6d2c3ef.png)

but it still is a bit awkward looking... (adding the error message and singular errors like this seems pretty hacky :confounded: )

what do you think @jessicajaniuk?

dario-piotrowicz replied:
no longer valid

dario-piotrowicz commented:
@jessicajaniuk added shared functions for both the errors throwing and the warnings logging, please have a look :smile:

jessicajaniuk commented on packages/animations/browser/src/errors-and-warnings.utils.ts:
```diff
@@ -0,0 +1,33 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+function generateErrorsOrWarningsCombination(message: string, errorsOrWarnings: string[]) {
```
So we've now strongly typed all of the errors as `Error` and they've been moved into a tree shaking friendly location. So this abstraction won't work for us. Take a look at the new code after a rebase. I think that'll give you a solid idea of the best way to make these warnings abstracted in a friendly and tree shakable way.

dario-piotrowicz replied:
ok sure, I'll have a look thanks :slightly_smiling_face::+1:

jessicajaniuk reviewed (changes requested):
@dario-piotrowicz I managed to break this again with the Error tree shaking pull request. So you'll want to take a look again. We'll want to be sure that warnings are also tree shakable, as well. Definitely take a look and let me know if you need any pointers. I'm pretty sure we can still get this in the v14 release.

dario-piotrowicz committed:
fix(animations): allow animations with unsupported CSS properties

currently animations with unsupported CSS properties cause a hard error
and the crash of the animation itself, instead of this behaviour just
ignore such properties and provide a warning for the developer in
the console (only in dev mode)

this change also introduces a general way to present warnings
in the animations code

resolves #23195

jessicajaniuk commented on packages/animations/browser/test/shared.ts:
```diff
@@ -13,16 +13,22 @@ import {buildAnimationAst} from '../src/dsl/animation_ast_builder';
   if (!skipErrors && errors.length) {
     throw triggerParsingFailed(name, errors);
   }
+  if ((typeof ngDevMode === 'undefined' || !!ngDevMode) && warnings.length) {
+    console.warn(triggerParsingWarnings(name, warnings));
+  }
```
Similar case here. You can simplify this since we're not shipping error codes and we're only warning in dev mode. The `console.warn` call can be embedded in the warnings helpers.

```suggestion
  if (warnings.length) {
    triggerParsingWarnings(name, warnings);
  }
```

jessicajaniuk commented on packages/animations/browser/src/warnings.ts:
```diff
@@ -0,0 +1,16 @@
+ * In production mode (after tree-shaking pass), the `message` argument becomes `false`,
+ * thus we account for it in the typings and the runtime logic.
+ */
+export type RuntimeWarning = string|false;
```
I think you can skip this. The reason we have a RuntimeError class is for special code handling and making it easy to manage errors / track them. Warnings don't have the same code system or needs. They're just strings. So we can just put those strings in the helpers class. It should still be tree shaken nicely.

jessicajaniuk commented on packages/animations/browser/src/render/animation_engine_next.ts:
```diff
@@ -44,11 +46,15 @@ export class AnimationEngine {
       if (errors.length) {
         throw triggerBuildFailed(name, errors);
       }
+      if ((typeof ngDevMode === 'undefined' || !!ngDevMode) && warnings.length) {
+        console.warn(triggerBuildWarnings(name, warnings));
+      }
```
```suggestion
      if (warnings.length) {
        triggerBuildWarnings(name, warnings);
      }
```

jessicajaniuk commented on packages/animations/browser/src/render/timeline_animation_engine.ts:
```diff
@@ -32,10 +34,14 @@ export class TimelineAnimationEngine {
     if (errors.length) {
       throw registerFailed(errors);
     } else {
+      if ((typeof ngDevMode === 'undefined' || !!ngDevMode) && warnings.length) {
+        console.warn(registerWarnings(warnings));
+      }
```
```suggestion
      if (warnings.length) {
        registerWarnings(warnings);
      }
```

jessicajaniuk reviewed (changes requested):
@dario-piotrowicz This is super close. I like the changes, and I have a few suggestions for making it a bit cleaner here. Should be very quick to address and we should be able to land it.

jessicajaniuk commented on packages/animations/browser/src/dsl/animation_ast_builder.ts:
```diff
@@ -56,22 +56,29 @@ const SELF_TOKEN_REGEX = new RegExp(`\s*${SELF_TOKEN}\s*,?`, 'g');
+
+    if (context.unsupportedCSSPropertiesFound.size) {
+      warnings.push(
+          `The provided CSS properties are not recognized properties supported for animations: ${
```
We should move this into a helper method, too. This will ensure the message itself gets shaken away. You could perhaps do one that just pushes to the warnings array. Something like the following in the helpers.

```
function warnUnrecognizedProperties(warnings: []string, props: []string): void {
  if (ngDevMode && props.length) {
     warnings.push(`[...your existing message goes here]`);
  }
}
```

dario-piotrowicz replied:
ah yeah sounds great! thanks! :smile::+1:

dario-piotrowicz replied:
done, I put the function in the warnings helper file, I hope that's ok :slightly_smiling_face:

jessicajaniuk commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -15,7 +15,7 @@
       "uncompressed": {
         "runtime": 4343,
-        "main": 449957,
+        "main": 450579,
```
I think you have a file conflict to resolve for this file.

dario-piotrowicz replied:
sorry I thought I solved it :sweat: :shrug:

jessicajaniuk reviewed (changes requested):
@dario-piotrowicz One last issue to resolve.

jessicajaniuk reviewed (approved):
LGTM :cookie:

Thanks for working through this, @dario-piotrowicz!

dario-piotrowicz commented:
> LGTM cookie
> 
> Thanks for working through this, @dario-piotrowicz!

Thank you for the awesome reviewing, sorry I went in the wrong direction a couple of times! :stuck_out_tongue_winking_eye:

alxhub commented:
@dario-piotrowicz currently this is tagged as `feat`, meaning it won't be mergeable into the `13.2.x` branch. However, this definitely feels like a bugfix instead. Should we update the commit tag and land this as a `fix`?

dario-piotrowicz commented:
> @dario-piotrowicz currently this is tagged as `feat`, meaning it won't be mergeable into the `13.2.x` branch. However, this definitely feels like a bugfix instead. Should we update the commit tag and land this as a `fix`?

Sorry I am not really sure.... the original code was __purposely__ stopping animations for unsupported properties, and I changed it to ignore them instead... so it does sort of sound like a feature (even if a very small one) :shrug: 

I don't even know what difference it makes to land this as a `feat` or a bugfix, I guess I don't mind either way, please let me know what you think/suggest

(@jessicajaniuk, what do you think? :slightly_smiling_face:)

dario-piotrowicz commented:
@alxhub Updated to `fix` :slightly_smiling_face::+1:

atscott commented:
FYI - This had a conflict with `13.2.x`. Please open a new PR to target that branch only.

atscott commented:
This PR was merged into the repository by commit f8dc66060591e1ce42265a7862c3df16e40b05e1.
